docker/release/config/opensearch/opensearch-docker-entrypoint.sh

```diff
@@ -90,22 +67,74 @@ function terminateProcesses {
     fi
 }
 
-# Enable job control so we receive SIGCHLD when a child process terminates
-set -m
+# Start up the opensearch and performance analyzer agent processes.
+# When either of them halts, this script exits, or we receive a SIGTERM or SIGINT signal then we want to kill both these processes.
+function runOpensearch {
+    # Files created by OpenSearch should always be group writable too
+    umask 0002
+
+    if [[ "$(id -u)" == "0" ]]; then
+        echo "OpenSearch cannot run as root. Please start your container as another user."
+        exit 1
+    fi
+
+    # Parse Docker env vars to customize OpenSearch
+    #
+    # e.g. Setting the env var cluster.name=testcluster
+    # will cause OpenSearch to be invoked with -Ecluster.name=testcluster
+    local opensearch_opts=()
+    while IFS='=' read -r envvar_key envvar_value
+    do
+        # OpenSearch settings need to have at least two dot separated lowercase
+        # words, e.g. `cluster.name`, except for `processors` which we handle
+        # specially
+        if [[ "$envvar_key" =~ ^[a-z0-9_]+\.[a-z0-9_]+ || "$envvar_key" == "processors" ]]; then
+            if [[ ! -z $envvar_value ]]; then
+            opensearch_opt="-E${envvar_key}=${envvar_value}"
+            opensearch_opts+=("${opensearch_opt}")
+            fi
+        fi
+    done < <(env)
+
+    setupSecurityPlugin
+
+    # Enable job control so we receive SIGCHLD when a child process terminates
+    set -m
 
-# Make sure we terminate the child processes in the event of us received TERM (e.g. "docker container stop"), INT (e.g. ctrl-C), EXIT (this script terminates for an unexpected reason), or CHLD (one of the processes terminated unexpectedly)
-trap terminateProcesses TERM INT EXIT CHLD
+    # Make sure we terminate the child processes in the event of us received TERM (e.g. "docker container stop"), INT (e.g. ctrl-C), EXIT (this script terminates for an unexpected reason), or CHLD (one of the processes terminated unexpectedly)
+    trap terminateProcesses TERM INT EXIT CHLD
 
-# Start opensearch
-$OPENSEARCH_HOME/bin/opensearch "${opensearch_opts[@]}" &
-OPENSEARCH_PID=$!
+    # Start opensearch
+    "$@" "${opensearch_opts[@]}" &
+    OPENSEARCH_PID=$!
 
-# Start performance analyzer agent
-$OPENSEARCH_HOME/bin/performance-analyzer-agent-cli > $OPENSEARCH_HOME/logs/performance-analyzer.log 2>&1 &
-PA_PID=$!
+    # Start performance analyzer agent
+    performance-analyzer-agent-cli > $OPENSEARCH_HOME/logs/performance-analyzer.log 2>&1 &
```

Comment:
Not sure if this is the best approach to rely on ` ENV PATH=$PATH:$OPENSEARCH_DASHBOARDS_HOME/bin` to find opensearch and pa cli script location.

I personally would like to give a full path to the script, but this is just nitpick tbh.

Reply:
The rationale was that I wanted the user to be able to specify `openseach` as a docker CMD, so having `opensearch` in PATH at least made sense for that reason. Then, as opensearch (and consequently PA) were looked up from PATH, why not also starting PA from PATH?

It'd still be possible to not add anything to PATH, and make the entrypoint "hijack" the "opensearch" command (which it already does it in a way) to run it using the absolute path, but I reckon it'd be needlessly convoluted.

Anyways, I don't have strong opinions on that either so I'll just let it up to you.

Reply:
Yeah, I get what you are saying.
I am not against this approach at all, just a bit questioning whether there are better ways.
Anyway, we can try this out for a few snapshots builds and see if it comes up.

https://hub.docker.com/r/opensearchstaging/opensearch
https://hub.docker.com/r/opensearchstaging/opensearch-dashboards